falcon/hooks.py

```diff
@@ -13,23 +13,41 @@
 # limitations under the License.
 
 """Hook decorators."""
+from __future__ import annotations
 
 from functools import wraps
 from inspect import getmembers
 from inspect import iscoroutinefunction
 import re
+from typing import Any
+from typing import Awaitable
+from typing import Callable
+from typing import Dict
+from typing import List
+from typing import Tuple
+from typing import TYPE_CHECKING
+from typing import Union
 
 from falcon.constants import COMBINED_METHODS
 from falcon.util.misc import get_argnames
 from falcon.util.sync import _wrap_non_coroutine_unsafe
 
+if TYPE_CHECKING:
+    import falcon as wsgi
+    from falcon import asgi
 
 _DECORABLE_METHOD_NAME = re.compile(
     r'^on_({})(_\w+)?$'.format('|'.join(method.lower() for method in COMBINED_METHODS))
 )
 
+SynchronousResource = Callable[..., Any]
+AsynchronousResource = Callable[..., Awaitable[Any]]
+Resource = Union[SynchronousResource, AsynchronousResource]
```

Comment:
I think we need to better distinguish between an action callable, a responder callable, and a resource class:

(a) An action callable is the the thing to execute before or after the designated HTTP method responders. It has a signature that is variadic but defining the type as a `typing.Protocol` might work, e.g.:

```py
class SynchronousAction(Protocol):
    def __call__(self, req: Request, resp: Response, resource: type, params: Mapping[str, str], *args: Any, **kwargs: Any) -> None: ...

Action = Union[SynchronousAction, AsynchronousAction]
```

(b) An HTTP method responder type, i.e. `func(self: type, req: Request, resp: Response, **kwargs: Any)`

(c) Resource class being the thing that contains HTTP method  `on_*()` responders. Since there is not base class for these, we can just use `type` to hint these.

Reply:
Added simple typing, will improve it in accordance to your remarks in PR: https://github.com/falconry/falcon/pull/2183